packages/gas-fee-controller/src/GasFeeControllerPolling.ts

```diff
@@ -0,0 +1,97 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseControllerV2 } from '@metamask/base-controller';
+import { safelyExecute } from '@metamask/controller-utils';
+import type { NetworkClientId } from '@metamask/network-controller';
+import type { Json } from '@metamask/utils';
+import { v1 as random } from 'uuid';
+
+export type PollingCompleteType<N extends string> = {
+  type: `${N}:pollingComplete`;
+  payload: [string];
+};
+
+export default abstract class GasFeeControllerPolling<
```

Comment:
Should this be added directly to the base controller instead?

Reply:
We wanted to separate this out to separate concerns and perhaps this could be the form the polling abstraction takes for other controllers to inherit this functionality.

Reply:
Yeah, this lets us add polling only to controllers with polling.

Using inheritance like this doesn't seem _ideal_. It wouldn't compose well if we have another "controller modification" type interface that we want to add on as well later. Better to use an interface if possible, which we can compose multiple of. But we can cross that bridge when we come to it

With an interface, the one thing I'm not sure of offhand is we get access to `this.messenger`.

Reply:
Agree, greater composability would be ideal, but if we used an interface we would have to duplicate the start, stop, stopAll and poll implementations in each controller though?

Reply:
Yes, but that's where we could use a library. We could move those implementations to `controller-utils` to reuse them, and use an interface for establishing a common interface for dealing with polling controllers

Reply:
Yep that makes sense

Reply:
could maybe use mixins pattern that typescript supports for this: https://www.typescriptlang.org/play#example/mixins

Reply:
Yeah, the mixin pattern looks applicable here. It seems that class decorators in TypeScript v5 and above might offer a more concise syntax for using the mixin pattern as well. But in either case, starting with an abstract class seems reasonable to me. It looks like even if we start using a mixin, the mixin could use this abstract class.